### PR TITLE
Add test command to abci-cli

### LINKED ISSUE #76: Test suite for evaluating implementations for bounty
There's some scripts that test eg. the "first abci guide" stuff. We should abstract this so it's easier to run on new implementations. Also, we should add an integration test to see it run with tendermint.

A successful implementation should come with a docker file so the server and app can be easily started. We can then run a series of txs and ensure things act as expected.

@adrianbrink want to work on this?

### PR DESCRIPTION
**Pull request by KrzysiekJ**
This supersedes #83 (commits squashed with slight changes + a simple fix) and fixes #76.

### CODE REVIEW

File: tests/server/client.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	res := client.CommitSync()
+	_, data, log := res.Code, res.Data, res.Log
+	if res.IsErr() {
+		fmt.Println("Failed test: Commit")
+		fmt.Printf("committing %v\nlog: %v", log)
+		return errors.New(res.Error())
+	}
+	if !bytes.Equal(data, hashExp) {
+		fmt.Println("Failed test: Commit")
+		fmt.Printf("Commit hash was unexpected. Got %X expected %X",
```
Comment (KrzysiekJ) [inline tests/server/client.go]: If this test fails, the first value is incorrectly rendered. For example, if the second value is `0000000000000005` and the first one is the same, but with 0 byte added at the end, we get “Got 303030303030303030303030303030353030 expected 0000000000000005”. Maybe it has something to do with the fact that the first value is `data.Bytes`, while the second is `[]byte`?

File: tests/server/client.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			data, hashExp)
+		return errors.New("CommitTx failed")
+	}
+	fmt.Println("Passed test: Commit")
+	return nil
+}
+
+func DeliverTx(client abcicli.Client, txBytes []byte, codeExp types.CodeType, dataExp []byte) error {
+	res := client.DeliverTxSync(txBytes)
```
Comment (melekes) [inline tests/server/client.go]: what about checking the `res.IsErr()`? similar to `CheckTx`

File: cmd/abci-cli/abci-cli.go [hunk 3]
```diff
@@ -196,6 +202,45 @@ func persistentArgs(line []byte) []string {
...
+	err = servertest.DeliverTx(client, []byte{0x00, 0x03}, types.CodeType_OK, nil)
+	fmt.Println("")
+	err = servertest.DeliverTx(client, []byte{0x00, 0x00, 0x04}, types.CodeType_OK, nil)
+	fmt.Println("")
+	err = servertest.DeliverTx(client, []byte{0x00, 0x00, 0x06}, types.CodeType_BadNonce, nil)
+	fmt.Println("")
+	err = servertest.Commit(client, []byte{0, 0, 0, 0, 0, 0, 0, 5})
+
+	if err != nil {
+		return errors.New("Some checks didn't pass, please use the cli to see the exact failures.")
```
Comment (melekes) [inline cmd/abci-cli/abci-cli.go]: huh? this command is a cli command. what do you mean by `use the cli`?

File: tests/server/client.go (hunk 1 above)
Comment (melekes) [reply tests/server/client.go]: have you tried `data.Bytes()`?

File: cmd/abci-cli/abci-cli.go [hunk 4]
```diff
@@ -143,14 +143,20 @@ func main() {
 				},
 			},
 		},
+		{
+			Name:  "test",
+			Usage: "Runs the integration tests",
+			Action: func(c *cli.Context) error {
+				return cmdTest(app, c)
+			},
+		},
```
Comment (zramsay) [inline cmd/abci-cli/abci-cli.go]: we're switching to cobra in #117, so this'll need to be re-worked. Maybe base this PR on that one?

File: cmd/abci-cli/abci-cli.go (hunk 3 above)
Comment (zramsay) [reply cmd/abci-cli/abci-cli.go]: i think he means "debug manually", like if in running `abci-cli test` there's a failure at one of the lines, use abci-cli deliver_tx` or whatever to debug. Instead, printing the actual `err` message on this line would show the exact failures, would it not?

File: tests/server/client.go (hunk 2 above)
Comment (KrzysiekJ) [reply tests/server/client.go]: Looks like `res.IsErr()` checks whether code is not `OK`. However we already have the `codeExp` argument. I guess that the check of `res.IsErr()` should be removed from `CheckTx` then, as it is redundant and will lead to mismatch if the caller expects something different than `OK`.

File: tests/server/client.go (hunk 1 above)
Comment (KrzysiekJ) [reply tests/server/client.go]: `data.Bytes()` helps, thanks.

File: cmd/abci-cli/abci-cli.go (hunk 4 above)
Comment (KrzysiekJ) [reply cmd/abci-cli/abci-cli.go]: I’ll rebase if #117 is merged earlier.

File: cmd/abci-cli/abci-cli.go (hunk 3 above)
Comment (KrzysiekJ) [reply cmd/abci-cli/abci-cli.go]: I’ll change “use the cli” to “inspect stdout”.

File: cmd/abci-cli/abci-cli.go [hunk 5]
```diff
@@ -196,6 +202,45 @@ func persistentArgs(line []byte) []string {
...
+	fmt.Println("")
+	err = servertest.DeliverTx(client, []byte{0x00, 0x03}, types.CodeType_OK, nil)
+	fmt.Println("")
+	err = servertest.DeliverTx(client, []byte{0x00, 0x00, 0x04}, types.CodeType_OK, nil)
+	fmt.Println("")
+	err = servertest.DeliverTx(client, []byte{0x00, 0x00, 0x06}, types.CodeType_BadNonce, nil)
+	fmt.Println("")
+	err = servertest.Commit(client, []byte{0, 0, 0, 0, 0, 0, 0, 5})
+
+	if err != nil {
```
Comment (KrzysiekJ) [inline cmd/abci-cli/abci-cli.go]: @zramsay: Haven’t noticed your comment (visible now in the outdated diff). It looks like there may be many errors, assigned in many checks and `err` in this line is only the last one (or `nil`), so I’m not sure whether previous assignments have any effect. Actually, if there are some errors, but the last commit test is correct, then it looks like this function will return `nil`… We should therefore probably change each `err = […]` assignment to something like `err = err || […]`.

Comment (KrzysiekJ) [reply cmd/abci-cli/abci-cli.go]: I’ve updated the commit, introducing a helper `or` function.

**Comment by melekes** [comment]
@KrzysiekJ sorry it took so long to respond. could you please rebase it against develop. let's merge this!

**Comment by KrzysiekJ** [comment]
Rebased. :)

File: README.md [hunk 6]
```diff
@@ -187,3 +187,47 @@ Here, we describe the requests and responses as function arguments and return va
...
+## In Process
+
+The simplest implementation just uses function calls within Go.
+This means ABCI applications written in Golang can be compiled with TendermintCore and run as a single binary.
+
+
+# Tools and Apps
+
+The `abci-cli` tool wraps any ABCI client and can be used for probing/testing an ABCI application.
+See the [guide](https://tendermint.com/docs/guides/abci-cli) for more details.
```
Comment (melekes) [inline README.md]: https://tendermint.readthedocs.io/en/master/abci-cli.html

File: Makefile [hunk 7]
```diff
@@ -8,8 +8,8 @@ all: protoc install test
 NOVENDOR = go list github.com/tendermint/abci/... | grep -v /vendor/
 
 install-protoc:
-	# Download: https://github.com/google/protobuf/releases
-	go get github.com/golang/protobuf/protoc-gen-go
+# Download: https://github.com/google/protobuf/releases
```
Comment (melekes) [inline Makefile]: ?

Comment (KrzysiekJ) [reply Makefile]: This was in the original PR. I’ve reverted that line (will push in a moment).

**Comment by KrzysiekJ** [comment]
I’ve resolved conflict with `develop` and updated the code to compile with recent changes. Hope it can be merged now.

**Comment by melekes** [comment]
`tests/test_cli/test.sh 100644 → 100755`

**Comment by melekes** [comment]
Thank you very much for working on this! Hopefully, @ebuchman will approve and merge it

File: cmd/abci-cli/abci-cli.go [hunk 8]
```diff
@@ -271,6 +274,16 @@ var dummyCmd = &cobra.Command{
 	},
 }
 
+var testCmd = &cobra.Command{
+	Use:   "test",
+	Short: "Run integration tests",
+	Long:  "",
```
Comment (zramsay) [inline cmd/abci-cli/abci-cli.go]: can we get a long description here? Even if it's the same as Short, they show up differently

File: cmd/abci-cli/abci-cli.go [hunk 9]
```diff
@@ -287,6 +300,53 @@ func persistentArgs(line []byte) []string {
...
+	err = or(err, servertest.DeliverTx(client, []byte{0x01}, code.CodeTypeOK, nil))
+	fmt.Println("")
+	err = or(err, servertest.DeliverTx(client, []byte{0x00, 0x02}, code.CodeTypeOK, nil))
+	fmt.Println("")
+	err = or(err, servertest.DeliverTx(client, []byte{0x00, 0x03}, code.CodeTypeOK, nil))
+	fmt.Println("")
+	err = or(err, servertest.DeliverTx(client, []byte{0x00, 0x00, 0x04}, code.CodeTypeOK, nil))
+	fmt.Println("")
+	err = or(err, servertest.DeliverTx(client, []byte{0x00, 0x00, 0x06}, code.CodeTypeBadNonce, nil))
+	fmt.Println("")
```
Comment (zramsay) [inline cmd/abci-cli/abci-cli.go]: can these `fmt.Println("")` be handled in `func or()` ?

Comment (melekes) [reply cmd/abci-cli/abci-cli.go]: I'd prefer not to.

File: cmd/abci-cli/abci-cli.go (hunk 8 above)
Comment (ebuchman) [reply cmd/abci-cli/abci-cli.go]: None of the others have it. Feel free to add them tho :)

**Pull request merged**